library/ssl_cli.c

```diff
@@ -757,17 +761,90 @@ static int ssl_validate_ciphersuite( const mbedtls_ssl_ciphersuite_t * suite_inf
     return( 0 );
 }
 
-static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
+/*
+ *
+ * STATE HANDLING: ClientHello
+ *
+ */
+
+/*
+ * Overview
+ */
+
+/* Main entry point; orchestrates the other functions */
+static int ssl_process_client_hello( mbedtls_ssl_context *ssl );
+
+static int ssl_client_hello_prepare( mbedtls_ssl_context *ssl );
+static int ssl_client_hello_write( mbedtls_ssl_context *ssl,
+                                   unsigned char* buf,
+                                   size_t buflen,
+                                   size_t *olen );
+static int ssl_client_hello_postprocess( mbedtls_ssl_context *ssl );
+
+/*
+ * Implementation
+ */
+
+static int ssl_process_client_hello( mbedtls_ssl_context *ssl )
 {
-    int ret;
-    size_t i, n, olen, ext_len = 0;
-    unsigned char *buf;
-    unsigned char *p, *q;
-    unsigned char offer_compress;
-    const int *ciphersuites;
-    const mbedtls_ssl_ciphersuite_t *ciphersuite_info;
+    int ret = 0;
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> process client hello" ) );
+
+    if( ssl->handshake->state_local.cli_hello_out.preparation_done == 0 )
+    {
+        MBEDTLS_SSL_PROC_CHK( ssl_client_hello_prepare( ssl ) );
```

Comment:
I would suggest moving the preparation_done = 1 to ssl_client_hello_prepare, or checking the return code of ssl_client_hello_prepare to not set the preparation_done upon encountering an error.

Reply:
The return code is checked through `MBEDTLS_SSL_PROC_CHK`. I prefer to keep the change of `preparation_done` here because this way the entire state change can be tracked from looking at the source of `ssl_process_client_hello`.